build_runner/test/integration_tests/optional_outputs_test.dart

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) 2018, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:build_test/build_test.dart';
+import 'package:meta/meta.dart';
+import 'package:test/test.dart';
+import 'package:test_descriptor/test_descriptor.dart' as d;
+
+import 'utils/build_descriptor.dart';
+
+// test-package-start
+/// Copies an asset to both `.txt.copy` and `.txt.extra`.
+final copyTwice = new TestBuilder(buildExtensions: {
+  '.txt': ['.txt.copy', '.txt.extra']
+});
+
+/// Reads `.txt.copy` files if the primary input contains "true".
+final maybeReadCopy = new TestBuilder(
+    buildExtensions: appendExtension('.other', from: '.txt'),
+    extraWork: (buildStep, _) async {
+      if ((await buildStep.readAsString(buildStep.inputId)).contains('true')) {
+        await buildStep.readAsString(buildStep.inputId.addExtension('.copy'));
+      }
+    });
+// test-package-end
+
+main() {
+  final builders = [
+    builder('copyTwice', copyTwice, isOptional: true),
+    builder('maybeReadCopy', maybeReadCopy, requiredInputs: ['.txt.copy'])
+  ];
+
+  BuildTool buildTool;
+
+  setUpAll(() async {
+    buildTool = await packageWithBuildScript(builders, contents: [
+      d.dir('web', [d.file('a.txt', 'false')])
+    ]);
+  });
+
+  Future<void> updateContent(String content) {
+    return d.dir('a', [
+      d.dir('web', [d.file('a.txt', content)])
+    ]).create();
+  }
+
+  group('serve', () {
+    test('only serves assets that were actually required', () async {
+      var server = await buildTool.serve();
+      await server.started;
+
+      await server.expect404('a.txt.copy');
+      await server.expect404('a.txt.extra');
+
+      await updateContent('true');
```

Comment:
nit: this feels a bit magic (update what content?). It isn't super bad since its defined in this file at least... maybe just `updateTxtContent`?

Reply:
Done